File: Objective-C/Objective-C/MyNativeBridge.m
//
//  MyNativeBridge.m
//  Objective-C
//
//  Copyright (c) 2015 DesktopJavaScript. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "MyNativeBridge.h"

@implementation MyNativeBridge

- (void)fetchMountedVolumes:(JSValue *)jsOptions {
    NSMutableArray *volumes = [self getMountedVolumes];
    
    JSValue *jsCallback = [jsOptions valueForProperty: @"callback"];
    [jsCallback callWithArguments:@[volumes]];
}

- (NSMutableArray *)getMountedVolumes {
    NSMutableArray *volumesList = [[NSMutableArray alloc] init];
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSArray *volumeResKeys = [NSArray arrayWithObjects:
                              NSURLVolumeLocalizedNameKey,
                              NSURLVolumeTotalCapacityKey,
                              NSURLVolumeAvailableCapacityKey,
                              NSURLVolumeIsBrowsableKey,
                              NSURLVolumeIdentifierKey,
                              NSURLVolumeURLKey,
                              nil];
    
    NSArray *volumeUrls = [fileManager mountedVolumeURLsIncludingResourceValuesForKeys: volumeResKeys options:0];
    for (NSURL *volUrl in volumeUrls) {
        NSError *volKeyError;
        NSString *volName;
        NSString *volIdentifer;
        NSNumber *volBrowsable;
        NSNumber *volAvailableCapacity;
        NSNumber *volTotalCapacity;
        
        
        [volUrl getResourceValue:&volName forKey:NSURLVolumeLocalizedNameKey error:&volKeyError];
        [volUrl getResourceValue:&volIdentifer forKey:NSURLVolumeURLKey error:&volKeyError];
        [volUrl getResourceValue:&volTotalCapacity forKey: NSURLVolumeTotalCapacityKey error:&volKeyError];
        [volUrl getResourceValue:&volAvailableCapacity forKey: NSURLVolumeAvailableCapacityKey error:&volKeyError];
        [volUrl getResourceValue:&volBrowsable forKey:NSURLVolumeIsBrowsableKey error:&volKeyError];
        
        NSByteCountFormatter *byteFormatter = [[NSByteCountFormatter alloc] init];
        byteFormatter.allowsNonnumericFormatting = NO;
        
        byteFormatter.includesUnit = NO;
        byteFormatter.includesCount = YES;
        NSString *availableCapacityCount = [byteFormatter stringFromByteCount:volAvailableCapacity.longLongValue];
        NSString *totalCapacityCount = [byteFormatter stringFromByteCount:volTotalCapacity.longLongValue];
        
        byteFormatter.includesUnit = YES;
        byteFormatter.includesCount = NO;
        NSString *availableCapacityUnit = [byteFormatter stringFromByteCount:volAvailableCapacity.longLongValue];
        NSString *totalCapacityUnit = [byteFormatter stringFromByteCount:volTotalCapacity.longLongValue];
        
        if ([volBrowsable boolValue]) {
            [volumesList addObject:[NSDictionary dictionaryWithObjectsAndKeys:
                                    volIdentifer, @"id",
                                    volName, @"name",
                                    availableCapacityCount, @"bytesAvailableCount",
                                    availableCapacityUnit, @"bytesAvailableUnit",
                                    totalCapacityCount, @"bytesTotalCount",
                                    totalCapacityUnit, @"bytesTotalUnit",
                                    nil]
             ];
        }
    }
    return volumesList;
}

@end

File: RubyMotion/app/my_native_bridge.rb
class MyNativeBridge < NativeBridge
  def fetchMountedVolumes(jsOptions)
    volumes = self.getMountedVolumes
    jsCallback = jsOptions["callback"]
    jsCallback.callWithArguments([volumes])
  end

  def getMountedVolumes
    volumesList = []
    fileManager = NSFileManager.defaultManager

    propertyKeys = [NSURLVolumeLocalizedNameKey,
                    NSURLVolumeTotalCapacityKey,
                    NSURLVolumeAvailableCapacityKey,
                    NSURLVolumeIsBrowsableKey,
                    NSURLVolumeIdentifierKey,
                    NSURLVolumeURLKey]
    volumeUrls = fileManager.mountedVolumeURLsIncludingResourceValuesForKeys(propertyKeys, options: 0)
    volumeUrls.each do |volumeUrl|
      volKeyError = Pointer.new(:object)
      volName = Pointer.new(:object)
      volIdentifier = Pointer.new(:object)
      volBrowsable = Pointer.new(:object)
      volAvailableCapacity = Pointer.new(:object)
      volTotalCapacity = Pointer.new(:object)

      volumeUrl.getResourceValue(volName, forKey: NSURLLocalizedNameKey, error: volKeyError)
      volumeUrl.getResourceValue(volIdentifier, forKey: NSURLVolumeIdentifierKey, error: volKeyError)
      volumeUrl.getResourceValue(volTotalCapacity, forKey: NSURLVolumeTotalCapacityKey, error: volKeyError)
      volumeUrl.getResourceValue(volAvailableCapacity, forKey: NSURLVolumeAvailableCapacityKey, error: volKeyError)
      volumeUrl.getResourceValue(volBrowsable, forKey: NSURLVolumeIsBrowsableKey, error: volKeyError)

      byteFormatter = NSByteCountFormatter.alloc.init
      byteFormatter.allowsNonnumericFormatting = false

      byteFormatter.includesUnit = false
      byteFormatter.includesCount = true
      availableCapacityCount = byteFormatter.stringFromByteCount(volAvailableCapacity[0])
      totalCapacityCount = byteFormatter.stringFromByteCount(volTotalCapacity[0])

      byteFormatter.includesUnit = true
      byteFormatter.includesCount = false
      availableCapacityUnit = byteFormatter.stringFromByteCount(volAvailableCapacity[0])
      totalCapacityUnit = byteFormatter.stringFromByteCount(volTotalCapacity[0])


      volumesList << {
        :id => volIdentifier[0],
        :name => volName[0],
        :bytesAvailableCount => availableCapacityCount,
        :bytesAvailableUnit  => availableCapacityUnit,
        :bytesTotalCount     => totalCapacityCount,
        :bytesTotalUnit      => totalCapacityUnit
      } if volBrowsable[0]
    end

    return volumesList
  end
end